Objects/object.c

```diff
@@ -2482,6 +2482,37 @@ _PyObject_SetDeferredRefcount(PyObject *op)
 #endif
 }
 
+int
+PyUnstable_Object_SetDeferredRefcount(PyObject *op)
+{
+#ifdef Py_GIL_DISABLED
+    if (!PyType_IS_GC(Py_TYPE(op)))
+    {
+        PyErr_SetString(PyExc_TypeError,
+                     "object is not tracked by the garbage collector");
+        return -1;
+    }
+
+    if (!_Py_IsOwnedByCurrentThread(op))
+    {
+        PyErr_SetString(PyExc_ValueError,
+                        "object is not owned by this thread");
+        return -1;
+    }
+
+    if (op->ob_ref_shared != 0)
+    {
+        PyErr_SetString(PyExc_ValueError,
+                        "object is already in use by another thread");
+        return -1;
+    }
+
+    _PyObject_SET_GC_BITS(op, _PyGC_BITS_DEFERRED);
+    op->ob_ref_shared = _Py_REF_SHARED(_Py_REF_DEFERRED, 0);
```

Comment:
I think we should consider trying to make this thread-safe rather than enforce the current refcount preconditions. These preconditions aren't exhaustive (they don't catch all non-thread-safe cases) and they're conservative (they disallow some thread-safe cases).

Basically:

* Set `_PyGC_BITS_DEFERRED` if it's not already set using an atomic operation. Either `_Py_atomic_or_uint8` or a compare-exchange loop.
* If `_PyGC_BITS_DEFERRED` was set, atomically add `_Py_REF_SHARED(_Py_REF_DEFERRED, 0)` to `op->ob_ref_shared`.

Reply:
I was under the impression that `_PyObject_SET_GC_BITS` was atomic already.